shade-test/jar-file/src/test/java/org/testcontainers/PublicAPITest.java

```diff
@@ -0,0 +1,120 @@
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.FieldNode;
+import org.objectweb.asm.tree.MethodNode;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+@RequiredArgsConstructor
+public class PublicAPITest extends AbstractJarFileTest {
+
+    private static String SHADED_PACKAGE = "org.testcontainers.shaded.";
+    private static String SHADED_PACKAGE_PATH = SHADED_PACKAGE.replaceAll("\\.", "/");
+
+    static {
+        Assertions.registerFormatterForType(ClassNode.class, it -> it.name);
+        Assertions.registerFormatterForType(FieldNode.class, it -> it.name);
+        Assertions.registerFormatterForType(MethodNode.class, it -> it.name + it.desc);
+    }
+
+    @Parameters(name = "{0}")
+    public static List<Object[]> data() throws Exception {
+        List<Object[]> result = new ArrayList<>();
+
+        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+
+            @Override
+            public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+                String fileName = path.toString();
+
+                if (!fileName.endsWith(".class")) {
+                    return super.visitFile(path, attrs);
+                }
+
+                if (!fileName.startsWith("/org/testcontainers/") ) {
+                     return super.visitFile(path, attrs);
+                }
+
+                if (fileName.startsWith("/" + SHADED_PACKAGE_PATH)) {
+                    return super.visitFile(path, attrs);
+                }
+
+                try(InputStream inputStream = Files.newInputStream(path)) {
+                    ClassReader reader = new ClassReader(inputStream);
+                    ClassNode node = new ClassNode();
+                    reader.accept(node, ClassReader.SKIP_CODE);
+                    if ((node.access & Opcodes.ACC_PUBLIC) != 0) {
+                        result.add(new Object[]{fileName, node});
+                    }
+                }
+
+                return super.visitFile(path, attrs);
+            }
+        });
+        return result;
+    }
+
+    private final String fileName;
+
+    private final ClassNode classNode;
+
+    @Test
+    public void testSuperClass() {
+        assertThat(classNode.superName)
+                .doesNotStartWith(SHADED_PACKAGE_PATH);
+    }
+
+    @Test
+    public void testInterfaces() {
+        assertThat(classNode.interfaces)
+                .allSatisfy(it -> assertThat(it).doesNotStartWith(SHADED_PACKAGE_PATH));
```

Comment:
We test final JAR, shaded. It might expose shaded classes - this is what we test.

It understand that the change might be confusing to you - it's a bit hardcore, but we need that to assure the quality of our project